Text/Earley/Derived.hs

```diff
@@ -30,3 +36,6 @@ list = listLike
 {-# INLINE listLike #-}
 listLike :: (Eq t, ListLike i t) => i -> Prod r e t i
 listLike = ListLike.foldr (liftA2 ListLike.cons . satisfy . (==)) (pure ListLike.empty)
+
+matches :: ListLike i t => (forall r. Grammar r (Prod r e t a)) -> i -> Bool
```

Comment:
Can you add documentation for this function?

Reply:
I'll add.
But I noticed that this isn't as lazy as it could be - ambiguous grammars such as
`mfix $ \p -> rule $ (terminal Just) <|> (p *> p)`
take an exponential time to decide using this function, but complete fast using `report`.
However, using `report` it's not possible to detect whether the empty word is accepted or rejected. So I'm not sure what's the best way to implement `matches` would be.

Reply:
Ah, I think this could have been implemented as a lower-level function by checking that a `Parsed` result is returned at the right position but without forcing the results, because I believe the list of results used to be guaranteed to be non-empty. But this might no longer be the case with this constraint validation feature. Hmm... :thinking:

Reply:
I'll try to think about it too, maybe there's some solution that covers a large subset of cases.
For now will merge as is.